src/Featurizers/Components/VectorNormsEstimator.h

```diff
@@ -0,0 +1,181 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+
+#include "TrainingOnlyEstimatorImpl.h"
+#include "../../Traits.h"
+#include "NormUpdaters.h"
+#include "../Structs.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+static constexpr char const * const         VectorNormsEstimatorName("VectorNormsEstimator");
+
+namespace TypeSelector {
+
+// updater type selector is created to select different updater based on input parameter
+template <typename T, NormType NormT>
+struct UpdaterTypeSelector {
+};
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::L1> {
+    using type = Updaters::L1NormUpdater<T>;
+};
+
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::L2> {
+    using type = Updaters::L2NormUpdater<T>;
+};
+
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::MAX> {
+    using type = Updaters::MaxNormUpdater<T>;
+};
+
+} // namespace TypeSelector
+
+
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         VectorNormsAnnotationData
+///  \brief         An annotation class which contains the norms of a matrix
+///
+class VectorNormsAnnotationData {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    std::vector<std::double_t> const Norms;
+    NormType                 const Type;

```

Comment:
>NormType                 const Type; [](start = 4, length = 36)

This should be a template rather than a data member, as we want to be able to store annotation data for both a L1 and L2 norm simultaneously. As implemented, get_annotation_data will return a cached L1 norm even when a L2 norm is requested (as it won't be able to resolve that those are 2 different things).